project/frontend/src/pages/Header.js

```diff
@@ -1,24 +1,105 @@
 import './Header.css';
 
-import React from 'react';
+import React, { useState, useEffect } from 'react';
+
+import Modal from '../components/Modal.js';
+import Loading from '../components/Loading.js';
 
 function Header(props) {
-  return (
-    <div id='header'>
-      {props.isUserLoggedIn ? (
-        <div>Pic</div>
-      ) : (
-        [
-          <button key='1' className='header-button'>
-            Create Account
-          </button>,
-          <button key='2' className='header-button'>
-            Log in
-          </button>,
-        ]
-      )}
-    </div>
-  );
+  let [GoogleAuth, setGoogleAuth] = useState(undefined);
+  let [currentUser, setCurrentUser] = useState(undefined);
+  let [signedIn, setSignedIn] = useState(false);
+  let [isModalOpen, setIsModalOpen] = useState(false);
+
+  useEffect(() => {
+    const onGoogleAuthLoaded = () => {
+      window.gapi.load('auth2', () => {
+        const auth2 = window.gapi.auth2.init({
+          client_id:
+            '21423804760-e3goj1cdhg49ojdf780mcq92qgshbr4v.apps.googleusercontent.com',
+        });
+
+        auth2.isSignedIn.listen((signedIn) => {
+          setSignedIn(signedIn);
+        });
+
+        auth2.currentUser.listen((user) => {
+          setCurrentUser(user);
+        });
+
+        setGoogleAuth(auth2);
+      });
+    };
+
+    if (window.gapi) {
+      onGoogleAuthLoaded();
+      return () => {};
+    } else {
+      window.addEventListener('DOMContentLoaded', onGoogleAuthLoaded);
+      return () => {
+        window.removeEventListener('DOMContentLoaded', onGoogleAuthLoaded);
+      };
+    }
+  }, []);
+
+  if (currentUser === undefined) {
+    return (
+      <div id='header'>
+        <Loading />
+      </div>
+    );
```

Comment:
Will this just be stuck in loading until the user signs in?

Reply:
The API is kind of weird, that's how I expected it to work as well, but after `window.gapi.auth2.init(...)` is called the user object is always initialized to an object regardless of whether a user is signed in. And if a user _is_ signed in (from a previous session or something) then `window.gapi.auth2.init(...)` actually signs them in automatically.

So basically the loading icon only shows up in the time after `DOMContentLoaded` and before `window.gapi.auth2.init(...)` is called, which in my experience is a fraction of a second.

Then, if a user is not logged in, `currentUser` will be assigned to some object that looks like `{wn: undefined, en: undefined}` and then when a user logs in, that object has its internal properties filled in.